src/serac/physics/utilities/quadrature_data.hpp

```diff
@@ -0,0 +1,197 @@
+    T*            data_ptr  = data_.data();
+    for (int i = 0; i < detail::retrieve(qfunc_).Size(); i += static_cast<int>(std::ceil(stride_))) {
+      // The only legal (portable, defined) way to do type punning in C++
+      std::memcpy(data_ptr + j, qfunc_ptr + i, sizeof(T));
+      j++;
+    }
+  }
+
+  // When a QuadratureData instance is managed by StateManager, we don't
+  // ever want to copy from or move from that instance.  This is sort of
+  // overkill when QuadratureData objects are created outside of the StateManager,
+  // but this case should be fairly rare.
+  QuadratureData(const QuadratureData&) = delete;
+  QuadratureData(QuadratureData&&)      = delete;
+
+  /**
+   * @brief Retrieves the data for a given quadrature point
+   * @param[in] element_idx The index of the desired element within the mesh
+   * @param[in] q_idx The index of the desired quadrature point within the element
+   */
+  T& operator()(const int element_idx, const int q_idx);
+
+  /**
+   * @brief Assigns an item to each quadrature point
+   * @param[in] item The item to assign
+   */
+  QuadratureData& operator=(const T& item);
+
+  /**
+   * @brief Iterator to the data for the first quadrature point
+   */
+  auto begin() { return data_.begin(); }
+  /// @overload
+  auto begin() const { return data_.begin(); }
+  /**
+   * @brief Iterator to one element past the data for the last quadrature point
+   */
+  auto end() { return data_.end(); }
+  /// @overload
+  auto end() const { return data_.end(); }
+
+  mfem::QuadratureFunction& QFunc() { return detail::retrieve(qfunc_); }
+
+  /**
+   * @brief Synchronizes data from the stored vector<T> to the raw double*
+   * array used by the underlying mfem::QuadratureFunction
+   *
+   * Used for saving to a file - MFEMSidreDataCollection
+   * (and by extension mfem::DataCollection's interface) only allow for
+   * quadrature-point-specific data via mfem::QuadratureFunction, so this logic
+   * is needed to glue together a generic array of data with that class
+   */
+  void sync() override
+  {
+    double*  qfunc_ptr = detail::retrieve(qfunc_).GetData();
+    int      j         = 0;
+    const T* data_ptr  = data_.data();
+    for (int i = 0; i < detail::retrieve(qfunc_).Size(); i += static_cast<int>(std::ceil(stride_))) {
+      // The only legal (portable, defined) way to do type punning in C++
+      std::memcpy(qfunc_ptr + i, data_ptr + j, sizeof(T));
+      j++;
+    }
+  }
+
+private:
+  /**
+   * @brief Storage layout of @p qfunc_ containing mesh and polynomial order info
+   */
+  detail::MaybeOwningPointer<mfem::QuadratureSpace> qspace_;
+  /**
+   * @brief Per-quadrature point data, stored as array of doubles for compatibility with Sidre
+   */
+  detail::MaybeOwningPointer<mfem::QuadratureFunction> qfunc_;
+
+  std::vector<T> data_;
+  /**
+   * @brief The stride of the array
+   */
+  static constexpr double stride_ = sizeof(T) / static_cast<double>(sizeof(double));
```

Comment:
I think making `stride_` an integer type, rather than double, would be more intuitive and make for cleaner code-- it looks like every use of `stride_` in the PR is of the form:
```cpp
static_cast<some_integer_type>(std::ceil(stride_))
```
which suggests that the quantity we actually care about is the integer value of the ceiling itself.

Reply:
Good suggestion, I had initially set it as a double before actually propagating it through the code but this definitely cleans things up!